### PR TITLE
Add SPI that allows extensions to listen on dev mode events

### PR DESCRIPTION
**Pull request by geoand**

### CODE REVIEW

File: core/deployment/src/main/java/io/quarkus/deployment/dev/IsolatedDevModeMain.java [hunk 1]
```diff
@@ -145,6 +146,16 @@ public void accept(String args) {
                 RuntimeUpdatesProcessor.INSTANCE.setConfiguredInstrumentationEnabled(
                         runner.getConfigValue("quarkus.live-reload.instrumentation", Boolean.class).orElse(false));
                 firstStartCompleted = true;
+
+                for (DevModeListener listener : ServiceLoader.load(DevModeListener.class)) {
```
Comment (gsmet) [inline core/deployment/src/main/java/io/quarkus/deployment/dev/IsolatedDevModeMain.java:150]: Naive question: we don't want to use build items for that?

**Comment by famod** [comment]
I guess some kind of ordering mechanism is out of scope for now and could be added later if people ask for it?

File: core/deployment/src/main/java/io/quarkus/deployment/dev/IsolatedDevModeMain.java (hunk 1 above)
Comment (geoand) [reply core/deployment/src/main/java/io/quarkus/deployment/dev/IsolatedDevModeMain.java:150]: They aren't available at this point. We might be able to make it work, but it's probably more work than it's worth.

**Comment by geoand** [comment]
> I guess some kind of ordering mechanism is out of scope for now and could be added later if people ask for it?

Added

File: core/deployment/src/main/java/io/quarkus/deployment/dev/IsolatedDevModeMain.java [hunk 2]
```diff
@@ -145,6 +147,25 @@ public void accept(String args) {
...
+                    listeners.add(listener);
+                }
+                listeners.sort(new Comparator<>() {
```
Comment (famod) [inline core/deployment/src/main/java/io/quarkus/deployment/dev/IsolatedDevModeMain.java:154]: Really minor:
```suggestion
                listeners.sort(Comparator.comparingInt(DevModeListener::order));
```
should do the same, no?

**Review by famod — APPROVED** [review]
LGTM (though I can't say I have a use case for this)

File: core/deployment/src/main/java/io/quarkus/deployment/dev/IsolatedDevModeMain.java (hunk 2 above)
Comment (geoand) [reply core/deployment/src/main/java/io/quarkus/deployment/dev/IsolatedDevModeMain.java:154]: Yeah, we can do that :(

**Comment by geoand** [comment]
The use case I have in mind is described at https://quarkusio.zulipchat.com/#narrow/stream/187038-dev/topic/ngrok/near/277111151

**Comment by geoand** [comment]
Although not a bug fix, it would be great to get this into `2.8` so I am adding the label

**Pull request merged**